arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Flow.kt

```diff
@@ -0,0 +1,104 @@
+ * The amount of concurrency depends on the
+ * [Characteristics] of [collector], and can
+ * be tweaked using the [concurrency] parameter.
+ *
+ * [this] is iterated only once during collection.
+ *
+ * Note: if you need to perform changes on the values
+ * before collection, we strongly recommend to convert
+ * the [Iterable] into a [Flow] using [asFlow],
+ * perform those changes, and then using [collect].
+ *
+ * @receiver Sequence of values to collect
+ * @param collector Describes how to collect the values
+ * @param concurrency Defines the concurrency limit
+ */
+@OptIn(FlowPreview::class)
+public suspend fun <T, R> Iterable<T>.collect(
+  collector: Collector<T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R = asFlow().collect(collector, concurrency)
+
+public fun <T, R> Iterable<T>.collectBlocking(
+  collector: Collector<T, R>,
+): R = asFlow().collectBlocking(collector)
+
+/**
+ * Performs collection over the elements of [this].
+ * The amount of concurrency depends on the
+ * [Characteristics] of [collector], and can
+ * be tweaked using the [concurrency] parameter.
+ *
+ * [this] is consumed only once during collection.
+ * We recommend using a cold [Flow] to ensure that
+ * elements are produced only when needed.
+ *
+ * @receiver [Flow] of elements to collect
+ * @param collector Describes how to collect the values
+ * @param concurrency Defines the concurrency limit
+ */
+@OptIn(FlowPreview::class)
+public suspend fun <T, R> Flow<T>.collect(
+  collector: Collector<T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R = collectI(collector, concurrency)
+
+@OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
+@Suppress("UNINITIALIZED_VARIABLE", "UNCHECKED_CAST")
+internal suspend fun <A, T, R> Flow<T>.collectI(
+  collector: CollectorI<A, T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R {
+  var result: A
+  val started = this.onStart { result = collector.supply() }
+  val continued = when {
+    Characteristics.CONCURRENT in collector.characteristics -> when {
+      Characteristics.UNORDERED in collector.characteristics ->
+        started.parMapUnordered(concurrency) { collector.accumulate(result, it) }
+
+      else ->
+        started.parMap(concurrency) { collector.accumulate(result, it) }
+    }
+
+    else -> started.map { collector.accumulate(result, it) }
+  }
+  var completed: R
+  continued.onCompletion {
+    completed = when {
+      Characteristics.IDENTITY_FINISH in collector.characteristics -> result as R
+      else -> collector.finish(result)
+    }
+  }.collect { }
+  return completed
+}
+
+@Suppress("UNINITIALIZED_VARIABLE")
+public fun <T, R> Flow<T>.collectBlocking(collector: Collector<T, R>): R {
+  var final: R
+  suspend { collect(collector) }.startCoroutine(object : Continuation<R> {
+    override val context: CoroutineContext = EmptyCoroutineContext
+    override fun resumeWith(result: Result<R>) {
+      final = result.getOrThrow()
+    }
+  })
+  return final
+}
```

Comment:
This is unsafe. Coroutines should never be ran with `EmptyCoroutineContext` (unless it's `@RestrictSuspension`, otherwise it breaks the semantics of context/thread preservation.

If anything within this `Flow` suspends, then `final` will result in an `NonInitializedException` since this is not actually blocking.

Only way to do this correctly is wrap it in `runBlocking`, but that's not available on JS... So I would personally omit this function in favor of `runBlocking { flow.collect(collector) }`.

Maybe worth adding one for `Sequence`, I think that would also be faster for `Iterable` to do `asSequence().collect(collector)` than `asFlow()` since `Sequence` has less overhead than `Flow` and runs "blocking".

Reply:
The problem with `Sequence` is that all the operations in `Collector` may `suspend` (although none of the basic collectors does). I'm wondering whether it's worth adding a subclass of `NonSuspendCollector`s that could run without the need of a `runBlocking` nor `suspend`.

Reply:
I've updated the code with `NonSuspend` variants which run on `Sequence` (or really on any `Iterator`). Now for `Iterable` you can use `collect` (non-suspend but no concurrency) or `parCollect` (suspend but uses concurrency if instructed).